lib/fun_with_flags/store/persistent/ecto.ex

```diff
@@ -134,6 +135,25 @@ defmodule FunWithFlags.Store.Persistent.Ecto do
     end
   end
 
+  # Is itself a transaction-wrapper function for SQLite3.
+  #
+  defp transaction_with_sqlite3(repo, upsert_fn) do
+    repo.transaction(fn ->
+      try do
+        upsert_fn.()
+      rescue
+        e ->
+          repo.rollback("Exception: #{inspect(e)}")
+      else
+        {:error, reason} ->
+          repo.rollback("Error while upserting the gate: #{inspect(reason)}")
+
+        {:ok, value} ->
+          {:ok, value}
+      end
+    end)
+  end
```

Comment:
I want to make sure I understand this.
Looking at [the SQLite transaction docs](https://www.sqlite.org/lang_transaction.html), a write transaction will automatically block other concurrent write transactions. Which makes sense because SQLite doesn't have MVCC:

> Transactions can be started manually using the BEGIN command. Such transactions usually persist until the next COMMIT or ROLLBACK command. But a transaction will also ROLLBACK if the database is closed or if an error occurs (...)
> SQLite supports multiple simultaneous read transactions coming from separate database connections, possibly in separate threads or processes, but only one simultaneous write transaction.
> A read transaction is used for reading only. A write transaction allows both reading and writing. A read transaction is started by a SELECT statement, and a write transaction is started by statements like CREATE, DELETE, DROP, INSERT, or UPDATE (collectively "write statements"). **If a write statement occurs while a read transaction is active, then the read transaction is upgraded to a write transaction** if possible. (...)
> (emphasis added)

The bold sentence would be relevant here because this is used to wrap a SELECT followed by an UPDATE.

And then:

> If some other database connection has already modified the database or is already in the process of modifying the database, then upgrading to a write transaction is not possible and the write statement will fail with [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy).

This is the important part for us, because that's the whole point of this transaction in Postgres and MySQL.

Reply:
Is the above correct?

Reply:
Yeah what you've written a more articulate version of my understanding. I wasn't sure whether it was still necessary so I opted to follow the pattern of the others in case. Do you think I should remove the extra `repo.transaction` there?

Reply:
Very late reply, but I think that we should keep the transaction block.

Everything should happen within a transaction in SQLite. The docs linked above say that transactions "can be started manually using the BEGIN command", which implies that it's not the only way to start them. Later the docs say that a simple `SELECT` starts a read transaction automatically.

Here we want to avoid race conditions between concurrent updates to the percentage gates (because we can't rely on the unique index, like we do for the other gates [1]).

I think that in the case of concurrent `SELECT+UPDATE`s, without transaction block the 4 operations (2 SELECTs, followed by 2 UPDATEs) could be interleaved in ways that could lead to writing stale data. So keeping the transaction block is necessary.

[1]: In hindsight this could be solved with a different table schema and a better index, but it's too late to do that without introducing a major breaking change.